<!DOCTYPE html>
<html class="no-js" lang="en">
<head>
	<meta charset="UTF-8">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<title>Promise.all和Promise.race的区别 - IT学习者博客</title>
	<script>(function(d,e){d[e]=d[e].replace("no-js","js");})(document.documentElement,"className");</script>
	<meta name="description" content="">
		<meta property="og:title" content="Promise.all和Promise.race的区别" />
<meta property="og:description" content="const f1 = new Promise(function (resolve, reject) { resolve(&#39;佳佳&#39;) }) const f2 = new Promise(function (resolve, reject) { reject(&#39;要&#39;) }) const f3 = new Promise(function (resolve, reject) { resolve(&#39;自律&#39;) }) const f4 = Promise.all([f1, f2, f3]) f4.then(function (value) { console.log(value, &#39;..成功&#39;) }).catch(function (error) { console.log(error, &#39;...失败&#39;) // 要 }) // 若f2 返回成功，那么这个value就返回一个数组[&#34;wo&#34;, &#34;ai&#34;, &#34;ni&#34;],但是f2返回的值被拒绝了，那么方法所返回的 Promise 就会立刻被 拒绝，而不必等待其他的 Promise 结束 const f5 = new Promise(function (resolve, reject) { setTimeout(() =&gt; { resolve(&#39;佳佳&#39;) },100) }) const f6 = new Promise(function (resolve, reject) { reject(&#39;要&#39;) }) const f7 = new Promise(function (resolve, reject) { resolve(&#39;自律&#39;) }) const f8 = Promise." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://itnewbiea.github.io/posts/3cb641339f1f78a33dc9397d0c0dd50b/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2020-08-19T21:30:34+08:00" />
<meta property="article:modified_time" content="2020-08-19T21:30:34+08:00" />


	<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
	<link rel="dns-prefetch" href="//fonts.googleapis.com">
	<link rel="dns-prefetch" href="//fonts.gstatic.com">
	<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:400,400i,700">

	<link rel="stylesheet" href="/css/style.css">
	

	<link rel="shortcut icon" href="/favicon.ico">
		
</head>
<body class="body">
	<div class="container container--outer">
		<header class="header">
	<div class="container header__container">
		
	<div class="logo">
		<a class="logo__link" href="/" title="IT学习者博客" rel="home">
			<div class="logo__item logo__text">
					<div class="logo__title">IT学习者博客</div>
					
				</div>
		</a>
	</div>
		<div class="divider"></div>
	</div>
</header>
		<div class="wrapper flex">
			<div class="primary">
			
<main class="main" role="main">
	<article class="post">
		<header class="post__header">
			<h1 class="post__title">Promise.all和Promise.race的区别</h1>
			
		</header>
		<div class="content post__content clearfix">
			
<div id="content_views" class="markdown_views prism-atom-one-light">
                    <svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
                        <path stroke-linecap="round" d="M5,0 0,2.5 5,5z" id="raphael-marker-block" style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"></path>
                    </svg>
                    <pre><code>        const f1 = new Promise(function (resolve, reject) {
            resolve('佳佳')
        })
        const f2 = new Promise(function (resolve, reject) {
            reject('要')
        })
        const f3 = new Promise(function (resolve, reject) {
            resolve('自律')
        })
        const f4 = Promise.all([f1, f2, f3])
        f4.then(function (value) {
            console.log(value, '..成功')
        }).catch(function (error) {
            console.log(error, '...失败')  // 要
        })

        // 若f2 返回成功，那么这个value就返回一个数组["wo", "ai", "ni"],但是f2返回的值被拒绝了，那么方法所返回的 Promise 就会立刻被 拒绝，而不必等待其他的 Promise 结束

        const f5 = new Promise(function (resolve, reject) {
            setTimeout(() =&gt; {
                resolve('佳佳')
            },100)
        })
        const f6 = new Promise(function (resolve, reject) {
            reject('要')
        })
        const f7 = new Promise(function (resolve, reject) {
            resolve('自律')
        })
        const f8 = Promise.race([f5, f6, f7])
        f8.then(function (success) {
            console.log(success, '..success') 
        }).catch(function (error) {
            console.log(error, '...失败') // 要
        })
        // Promise.race([f5, f6, f7]) 调用顺序是 5，6，7，按说应该先执行f5，但是在f5执行的时候加了一个延时，这样他会加入到异步队列去执行，这样执行到f6到时候，返回的值被拒绝了，那么返回值就是这个被拒绝的值，

        // all和race区别：
        // race 翻译过来就是赛跑的意思，意味着Promise.race([f2, f1, f3])，这里面哪个参数执行的快，就返回哪一个，它不会在意返回的结果
        // all  成功的时候返回一个成功数组，只要有一个失败，则返回最先被reject失败状态的值
</code></pre>
                </div>
		</div>
	</article>
</main>


<nav class="pager flex">
	<div class="pager__item pager__item--prev">
		<a class="pager__link" href="/posts/e2a80038d4ed785c33204b1a2602d412/" rel="prev">
			<span class="pager__subtitle">«&thinsp;Previous</span>
			<p class="pager__title">[护网杯 2018]easy_tornado 1（STTI模板注入&#43;Tornado的secret_cookie）</p>
		</a>
	</div>
	<div class="pager__item pager__item--next">
		<a class="pager__link" href="/posts/baf63b75c17a3547d016fc347806b764/" rel="next">
			<span class="pager__subtitle">Next&thinsp;»</span>
			<p class="pager__title">亲测-windows系统安装Tensorflow2.0版本</p>
		</a>
	</div>
</nav>


			</div>
			
		</div>
		<footer class="footer">
	<div class="container footer__container flex">
		
		<div class="footer__copyright">
			&copy; 2024 IT学习者博客.
			<span class="footer__copyright-credits">Generated with <a href="https://gohugo.io/" rel="nofollow noopener" target="_blank">Hugo</a> and <a href="https://github.com/Vimux/Mainroad/" rel="nofollow noopener" target="_blank">Mainroad</a> theme.</span>
		</div>
	</div>
</footer>
<script src="https://itnewbiea.github.io/js/foot.js"></script>


	</div>
<script async defer src="/js/menu.js"></script>
</body>
</html>